astropy/io/ascii/ui.py

```diff
@@ -618,37 +616,23 @@ def _get_guess_kwargs_list(read_kwargs):
     # (actually include all here - check for compatibility of arguments later).
     # FixedWidthTwoLine would also be read by Basic, so it needs to come first;
     # same for RST.
-    for reader in [fixedwidth.FixedWidthTwoLine, rst.RST,
+    for reader in (fixedwidth.FixedWidthTwoLine, rst.RST,
                    fastbasic.FastBasic, basic.Basic,
                    fastbasic.FastRdb, basic.Rdb,
                    fastbasic.FastTab, basic.Tab,
                    cds.Cds, daophot.Daophot, sextractor.SExtractor,
-                   ipac.Ipac, latex.Latex, latex.AASTex]:
-        first_kwargs = copy.deepcopy(read_kwargs)
-        first_kwargs.update(dict(Reader=reader))
-        guess_kwargs_list.append(first_kwargs)
+                   ipac.Ipac, latex.Latex, latex.AASTex):
+        guess_kwargs_list.append(dict(Reader=reader))
 
     # Cycle through the basic-style readers using all combinations of delimiter
-    # and quotechar, unless already specified explicitly in input.
-    # The latter might result in duplicate attempts for [Fast]Basic readers...
-    for reader in [fastbasic.FastCommentedHeader, basic.CommentedHeader,
+    # and quotechar.
+    for Reader in (fastbasic.FastCommentedHeader, basic.CommentedHeader,
```

Comment:
Note - I'd like to revert to the original code / logic here.  The design is to largely decouple this code (generating the full guess list) as much from the user kwargs.  The filtering and de-duplication is done upstream the guessing code.  

Note also that user kwargs are added in the filter process, so your change to remove the original first few guesses with kwargs explicitly added was a good one and simplifies the code.

Reply:
For the default options this seems to work as intended, but with `fast_reader='force'` now the actual guess sequence is changed:
`3.0.1`:
```
>>> t1 = ascii.read('a\tb\n1\t2', fast_reader=True)
>>> for k in ascii.get_read_trace(): print("{:} - {:}".format(k.get('kwargs').get('Reader'), k.get('status')))
... 
<class 'astropy.io.ascii.ecsv.Ecsv'> - InconsistentTableError: ECSV header line like "# %ECSV <version>" not found as first line.  This is required for a ECSV file.
<class 'astropy.io.ascii.fixedwidth.FixedWidthTwoLine'> - InconsistentTableError: Position line should only contain delimiters and one other character, e.g. "--- ------- ---".
<class 'astropy.io.ascii.fastbasic.FastBasic'> - ValueError: Table format guessing requires at least two columns, got ['a\tb']
<class 'astropy.io.ascii.basic.Basic'> - ValueError: Table format guessing requires at least two columns, got ['a\tb']
<class 'astropy.io.ascii.basic.Rdb'> - ValueError: RDB types definitions do not all match [num](N|S): ['1', '2']
<class 'astropy.io.ascii.fastbasic.FastTab'> - Success (guessing)
>>> t1 = ascii.read('a\tb\n1\t2', fast_reader='force')
>>> for k in ascii.get_read_trace(): print("{:} - {:}".format(k.get('kwargs').get('Reader'), k.get('status')))
... 
<class 'astropy.io.ascii.fastbasic.FastBasic'> - ValueError: Table format guessing requires at least two columns, got ['a\tb']
<class 'astropy.io.ascii.fastbasic.FastTab'> - Success (guessing)
```
This branch:
```
>>> t2 = ascii.read('a\tb\n1\t2', fast_reader=True)
>>> for k in ascii.get_read_trace(): print("{:} - {:}".format(k.get('kwargs').get('Reader'), k.get('status')))
... 
<class 'astropy.io.ascii.ecsv.Ecsv'> - InconsistentTableError: ECSV header line like "# %ECSV <version>" not found as first line.  This is required for a ECSV file.
<class 'astropy.io.ascii.fixedwidth.FixedWidthTwoLine'> - InconsistentTableError: Position line should only contain delimiters and one other character, e.g. "--- ------- ---".
<class 'astropy.io.ascii.rst.RST'> - InconsistentTableError: Position line should only contain delimiters and one other character, e.g. "--- ------- ---".
<class 'astropy.io.ascii.fastbasic.FastBasic'> - ValueError: Table format guessing requires at least two columns, got ['a\tb']
<class 'astropy.io.ascii.basic.Basic'> - ValueError: Table format guessing requires at least two columns, got ['a\tb']
<class 'astropy.io.ascii.fastbasic.FastRdb'> - ValueError: RDB type definitions do not all match [num](N|S): ['1', '2']
<class 'astropy.io.ascii.basic.Rdb'> - ValueError: RDB types definitions do not all match [num](N|S): ['1', '2']
<class 'astropy.io.ascii.fastbasic.FastTab'> - Success (guessing)
>>> t2 = ascii.read('a\tb\n1\t2', fast_reader='force')
>>> for k in ascii.get_read_trace(): print("{:} - {:}".format(k.get('kwargs').get('Reader'), k.get('status')))
... 
<class 'astropy.io.ascii.ecsv.Ecsv'> - No fast version of reader available
<class 'astropy.io.ascii.fixedwidth.FixedWidthTwoLine'> - No fast version of reader available
<class 'astropy.io.ascii.rst.RST'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Rdb'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Tab'> - No fast version of reader available
<class 'astropy.io.ascii.cds.Cds'> - No fast version of reader available
<class 'astropy.io.ascii.daophot.Daophot'> - No fast version of reader available
<class 'astropy.io.ascii.sextractor.SExtractor'> - No fast version of reader available
<class 'astropy.io.ascii.ipac.Ipac'> - No fast version of reader available
<class 'astropy.io.ascii.latex.Latex'> - No fast version of reader available
<class 'astropy.io.ascii.latex.AASTex'> - No fast version of reader available
<class 'astropy.io.ascii.basic.CommentedHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.CommentedHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.CommentedHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.CommentedHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.CommentedHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.CommentedHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.CommentedHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.CommentedHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.Basic'> - No fast version of reader available
<class 'astropy.io.ascii.basic.NoHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.NoHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.NoHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.NoHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.NoHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.NoHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.NoHeader'> - No fast version of reader available
<class 'astropy.io.ascii.basic.NoHeader'> - No fast version of reader available
<class 'astropy.io.ascii.fastbasic.FastBasic'> - ValueError: Table format guessing requires at least two columns, got ['a\tb']
<class 'astropy.io.ascii.fastbasic.FastRdb'> - ValueError: RDB type definitions do not all match [num](N|S): ['1', '2']
<class 'astropy.io.ascii.fastbasic.FastTab'> - Success (guessing)
```
The result is the same as the guessing eventually proceeds to `fastbasic.FastTab` and returning the `basic.*` versions with a `No fast version of reader available` status instead of skipping seems OK, but I think we still want to maintain the original order of `fastbasic`/`basic` readers.

Reply:
OK, this is a spurious trace since it now lists all the readers passed over in the assembly of `filtered_guess_kwargs` - the actually tried readers are only the last 3.